### PR TITLE
Make cpp assembly file extensions case sensitive again

### PR DESCRIPTION
**Pull request by Gormo**
This fixes an issue introduced by PR #14005 where .s and .S
extensions were handled case-insensitive on Windows so the action
assemble was triggered instead of preprocess_assemble.

### CODE REVIEW

**Comment by Gormo** [comment]
@lberki, can we merge this so we don't need to maintain this patch locally when doing uplifts?

**Comment by oquenchil** [comment]
@buildbreaker2021 will take over the review and import.

**Review by buildbreaker2021 — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppFileTypes.java [hunk 1]
```diff
@@ -91,15 +91,43 @@ public boolean apply(String path) {
         }
       };
 
-  public static final FileType ASSEMBLER_WITH_C_PREPROCESSOR = FileType.of(".S");
-  public static final FileType PIC_ASSEMBLER = FileType.of(".pic.s");
+  public static final FileType ASSEMBLER_WITH_C_PREPROCESSOR =
```
[the remark this answers is not in the document]
Comment (Gormo) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CppFileTypes.java]: Good point.. I've reverted that part.

File: src/main/java/com/google/devtools/build/lib/rules/cpp/CppFileTypes.java [hunk 2]
```diff
@@ -91,15 +91,43 @@ public boolean apply(String path) {
...
+          return path.endsWith(ext);
+        }
+
+        @Override
+        public ImmutableList<String> getExtensions() {
+          return ImmutableList.of(ext);
+        }
+      };
+
+  public static final FileType PIC_ASSEMBLER =
```
[the remark this answers is not in the document]
Comment (Gormo) [reply src/main/java/com/google/devtools/build/lib/rules/cpp/CppFileTypes.java:95]: Test added!

**Review by buildbreaker2021 — APPROVED** [review]
lgtm

**Review by comius — CHANGES_REQUESTED** [review]
This PR seems to negate the assumptions made in WindowsOsPathPolicy.endsWith, that Windows don't know about character case (upper/lower). https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/vfs/WindowsOsPathPolicy.java;l=213;drc=9cb59369a84d9328c2929eefcb58c1b972ab15f2

Having a special case for Windows in CppFileTypes introduces a tech debt in my opinion - there's lack of comments (in this PR) why this there is a need for a special case here. Please add a comment with a link to an issue.

I think the right thing to do is to remove the assumption (of case insensitivity) in WindowsOsPathPolicy and then fix the locations where poorly behaved toolchains create either lower or upper cased suffices.

**Comment by torgil** [comment]
> I think the right thing to do is to remove the assumption (of case insensitivity) in WindowsOsPathPolicy and then fix the locations where poorly behaved toolchains create either lower or upper cased suffices.

This would reintroduce the issues in #14005. The "right thing" question in there was never properly discussed but you could reason that a new windows user would have a hard time to "fix" third-party libraries for casing (Microsoft clibs package in my case) or make Windows-specific case-by-case addition to CppFileTypes.

**Comment by comius** [comment]
> > I think the right thing to do is to remove the assumption (of case insensitivity) in WindowsOsPathPolicy and then fix the locations where poorly behaved toolchains create either lower or upper cased suffices.
> 
> This would reintroduce the issues in #14005. The "right thing" question in there was never properly discussed but you could reason that a new windows user would have a hard time to "fix" third-party libraries for casing (Microsoft clibs package in my case) or make Windows-specific case-by-case additions to CppFileTypes.

Sure, I agree.

Would you mind just adding a comment to the code, that "FileType is extended to use case-sensitive comparison also on Windows". I don't want the next developer hitting their head on it.

When this piece of code is rewritten to Starlark, we won't have the luxury of extending FileType, all file types will need to be listed case-sensitively. Perhaps special cased for linux/windows/macos one level higher.

I'm not sure if it's a good or a bad news we only need to special case this one special operating system :)

**Review by comius — APPROVED** [review]

**Comment by Gormo** [comment]
Ok for merge now?

**Comment by fmeum** [comment]
@bazel-io flag

**Pull request closed without merging**